File: src/MCPhoto.Core/Models/UserRole.cs
namespace MCPhoto.Core.Models;

/// <summary>계정 역할. (PRD §F8, it13 TempUser, it16 AdvancedUser)</summary>
public enum UserRole
{
    // ⚠️ 서수(enum 값)를 위계 비교에 쓰지 않는다(CanManage는 ManageRank switch). 배치값은 가독성용으로 위계 순 명시.
    //    저장·전송은 전부 ToFirestoreValue() 문자열이라 배치값 변경은 무해(it13 §1.1·§3.1).
    //    it16: AdvancedUser를 위계 순 위치(2)에 끼워 넣고 Manager·Admin을 3·4로 밀었다(설계 §3.1 안전 게이트 통과).

    /// <summary>임시 유저(it13): user와 동기능 + QR 전송(업로드+다운로드)만 시간·횟수 한도. 위계 최하위.</summary>
    TempUser = 0,

    /// <summary>AppSettings 관리. it16부터 프레임은 **사용만**(생성·편집·삭제 불가, E4).</summary>
    User = 1,

    /// <summary>고급 유저(it16): User 권한 + 프레임 생성·편집·삭제(개인 로컬). power 아님(계정 관리 권한 없음).</summary>
    AdvancedUser = 2,

    /// <summary>advanced_user + 사용자 관리 + 공용 기본 프레임 관리.</summary>
    Manager = 3,

    /// <summary>manager + manager 지정(최종 1인).</summary>
    Admin = 4
}

/// <summary>UserRole 문자열 매핑(Firestore 저장값과 일치).</summary>
public static class UserRoleExtensions
{
    public static string ToFirestoreValue(this UserRole role) => role switch
    {
        UserRole.TempUser => "temp_user",   // it13: C#↔TS↔Firestore 일관(snake_case)
        UserRole.User => "user",
        UserRole.AdvancedUser => "advanced_user",   // it16 §5.1 동결표
        UserRole.Manager => "manager",
        UserRole.Admin => "admin",
        _ => "user"
    };

    public static UserRole ParseRole(string? value) => value switch
    {
        "admin" => UserRole.Admin,
        "manager" => UserRole.Manager,
        "advanced_user" => UserRole.AdvancedUser,   // it16
        "temp_user" => UserRole.TempUser,   // it13
        "user" => UserRole.User,            // 명시(기존 default 폴백에서 승격)
        _ => UserRole.User                  // 미지원값 폴백 유지(오탈자 시 최소권한 — it16부터 프레임 쓰기도 없어 더 안전)
    };

    /// <summary>
    /// power 계정(사용자 관리·공용 기본 프레임 관리 권한).
    /// ⚠️ it16: AdvancedUser는 **여기에 포함되지 않는다**(계정 관리 권한 없음). 프레임 저작 권한은
    ///    별개 축인 <see cref="CanWriteFrames"/>다 — 두 판정을 서로 대체하지 않는다.
    /// </summary>
    public static bool IsPower(this UserRole role) => role is UserRole.Manager or UserRole.Admin;

    /// <summary>
    /// 프레임 쓰기 권한(생성·편집·삭제). AdvancedUser 이상. (it16 E2)
    /// ⚠️ <see cref="IsPower"/>와 **별개 축**이다: IsPower=계정 관리·공용 DB 프레임 관리, CanWriteFrames=프레임 저작.
    ///    AdvancedUser는 CanWriteFrames=true, IsPower=false다. 두 판정을 서로 대체하지 않는다.
    ///    <see cref="ManageRank"/> 부등식으로 쓰지 않는다 — 관리 위계에 역할이 끼어들 때 저작 권한이
    ///    조용히 따라 움직이는 것을 막기 위해 명시 열거를 유지한다(it13이 서수를 버린 것과 같은 이유).
    /// </summary>
    public static bool CanWriteFrames(this UserRole role)
        => role is UserRole.AdvancedUser or UserRole.Manager or UserRole.Admin;

    /// <summary>역할 한글 표시 라벨(계정 생성 콤보·사용자 관리 목록·팝오버 등, it13 §9.1). 미지원값은 "사용자".</summary>
    public static string ToLabel(this UserRole role) => role switch
    {
        UserRole.TempUser => "임시 유저",
        UserRole.User => "사용자",
        UserRole.AdvancedUser => "고급 유저",   // it16
        UserRole.Manager => "매니저",
        UserRole.Admin => "관리자",
        _ => "사용자"
    };

    /// <summary>
    /// actingRole이 생성할 수 있는 역할 목록(it2 §7, it13 §3.1, it16 §3.6):
    /// admin→[TempUser,User,AdvancedUser,Manager], manager→[TempUser,User,AdvancedUser], 그 외→[].
    /// (admin→admin 불가: 최종 1인 규칙)
    /// ⚠️ it15의 계정 생성 폐지로 프로덕션 호출자가 0이다(테스트만 참조). 삭제하지 않고 목록만 §3.3 매트릭스와
    ///    맞춰 둔다 — 훗날 되살아날 때 E3와 모순되는 규칙이 조용히 부활하는 것을 막는다.
    /// </summary>
    public static IReadOnlyList<UserRole> CreatableRoles(this UserRole actingRole) => actingRole switch
    {
        UserRole.Admin => new[] { UserRole.TempUser, UserRole.User, UserRole.AdvancedUser, UserRole.Manager },
        UserRole.Manager => new[] { UserRole.TempUser, UserRole.User, UserRole.AdvancedUser },
        _ => Array.Empty<UserRole>()
    };

    /// <summary>actingRole이 role 계정을 생성할 권한이 있는지(게이트 판정).</summary>
    public static bool CanCreate(this UserRole actingRole, UserRole role)
        => actingRole.CreatableRoles().Contains(role);

    /// <summary>
    /// 위계 랭크(관리 판정·표시 정렬 기준). 서수(enum 값)와 분리해 명시 — 역할 추가 시 여기만 갱신(서수 재배치 안전, it13 §3.2).
    /// 위계: TempUser &lt; User &lt; AdvancedUser &lt; Manager &lt; Admin. 서버 MANAGE_RANK와 동일(it16 §5.1).
    /// ⚠️ 권한 판정에 이 값을 직접 부등식으로 쓰지 않는다 — 관리 가능 여부는 <see cref="CanManage"/>,
    ///    프레임 저작은 <see cref="CanWriteFrames"/>다. 여기서 공개하는 목적은 **목록 정렬**(사용자 관리 화면)이다.
    /// </summary>
    public static int HierarchyRank(this UserRole role) => role switch
    {
        UserRole.TempUser => 0,
        UserRole.User => 1,
        UserRole.AdvancedUser => 2,   // it16
        UserRole.Manager => 3,
        UserRole.Admin => 4,
        _ => 0
    };

    /// <summary>위계 랭크(내부 별칭 — 기존 관리 판정 표현 유지).</summary>
    private static int ManageRank(UserRole role) => role.HierarchyRank();

    /// <summary>
    /// actingRole이 targetRole 계정을 관리(삭제·PIN 재설정 등)할 수 있는지: **자신과 같거나 낮은 위계만**.
    /// 위계 TempUser&lt;User&lt;AdvancedUser&lt;Manager&lt;Admin. 예) manager는 admin을 관리 불가, admin은 전부 관리 가능.
    /// ⚠️ 서수 대소 비교가 아니라 <see cref="ManageRank"/> 명시 랭크로 판정(향후 역할 추가에도 안전).
    /// ⚠️ 이 판정만으로는 비power도 통과한다 — 관리 액션 게이트는 <see cref="IsPower"/>와 **함께** 쓴다(it16 §3.5).
    /// </summary>
    public static bool CanManage(this UserRole actingRole, UserRole targetRole)
        => ManageRank(targetRole) <= ManageRank(actingRole);
}

File: webclient/src/domain/roles/userRole.ts
/**
 * 계정 역할 — Windows `Models/UserRole.cs` 이식 (analysis/60 §1)
 *
 * ⚠️ **문자열 유니온을 진실값으로 쓴다.** Windows는 enum + `ToFirestoreValue()`로 나뉘어 있지만
 *    웹은 서버·Firestore·로그가 쓰는 snake_case 문자열이 곧 값이다(직렬화 경계가 없어 드리프트 불가).
 * ⚠️ **위계 비교에 배열 순서를 쓰지 않는다.** `hierarchyRank`가 유일한 랭크 정의다.
 */

export const USER_ROLES = ["temp_user", "user", "advanced_user", "manager", "admin"] as const;
export type UserRole = (typeof USER_ROLES)[number];

/** 미지원·오탈자 값의 폴백(최소 권한). */
export const DEFAULT_ROLE: UserRole = "user";

/** 서버·저장값 문자열 → 역할. 알 수 없는 값은 최소 권한으로 폴백한다. */
export function parseRole(value: string | null | undefined): UserRole {
  if (typeof value !== "string") return DEFAULT_ROLE;
  return (USER_ROLES as readonly string[]).includes(value) ? (value as UserRole) : DEFAULT_ROLE;
}

/** 역할 한글 표시 라벨(analysis/13 §14 카탈로그). */
export function roleLabel(role: UserRole): string {
  switch (role) {
    case "temp_user":
      return "임시 유저";
    case "user":
      return "사용자";
    case "advanced_user":
      return "고급 유저";
    case "manager":
      return "매니저";
    case "admin":
      return "관리자";
    default:
      return "사용자";
  }
}

/**
 * 위계 랭크. 서버 `MANAGE_RANK`와 동일해야 한다(analysis/60 §1).
 * 공개 목적은 **목록 정렬**이다 — 권한 판정에 부등식으로 직접 쓰지 않는다.
 */
export function hierarchyRank(role: UserRole): number {
  switch (role) {
    case "temp_user":
      return 0;
    case "user":
      return 1;
    case "advanced_user":
      return 2;
    case "manager":
      return 3;
    case "admin":
      return 4;
    default:
      return 0;
  }
}

/**
 * power 계정(사용자 관리 · 공용 기본 프레임 관리).
 * ⚠️ `advanced_user`는 **포함되지 않는다** — 프레임 저작 권한은 별개 축인 `canWriteFrames`다.
 */
export function isPower(role: UserRole): boolean {
  return role === "manager" || role === "admin";
}

/**
 * 프레임 쓰기 권한(생성·편집·삭제). `advanced_user` 이상.
 * ⚠️ `isPower`와 **별개 축**이다. `hierarchyRank` 부등식으로 쓰지 않는다 — 관리 위계에 역할이
 *    끼어들 때 저작 권한이 조용히 따라 움직이는 것을 막기 위해 명시 열거를 유지한다.
 */
export function canWriteFrames(role: UserRole): boolean {
  return role === "advanced_user" || role === "manager" || role === "admin";
}

/**
 * 관리(삭제 등) 가능 여부: **자신과 같거나 낮은 위계**.
 * ⚠️ 이 판정만으로는 비power도 통과한다 — 관리 액션 게이트는 `isPower`와 **함께** 쓴다.
 */
export function canManage(actingRole: UserRole, targetRole: UserRole): boolean {
  return hierarchyRank(targetRole) <= hierarchyRank(actingRole);
}

/**
 * 타 계정 PIN 재설정 권한. 관리 액션 중 **유일하게 "엄격히 낮은 위계"**만 허용한다.
 * power + 대상 위계가 자신보다 낮을 때만: admin→manager ○, manager→manager ✕(admin 전용).
 *
 * 왜 `canManage`가 아닌가: PIN은 설정·계정 진입의 유일한 자격증명이므로 동급끼리 서로의 진입 자격을
 * 갈아치우는 것은 과대 권한이다. 서버 `domain/roles.ts canResetPin`과 1:1이다.
 */
export function canResetPin(actingRole: UserRole, targetRole: UserRole): boolean {
  return isPower(actingRole) && hierarchyRank(targetRole) < hierarchyRank(actingRole);
}

/**
 * 생성 가능한 역할 목록(analysis/60 §1.4).
 * ⚠️ it15의 계정 생성 폐지로 프로덕션 호출자가 없다 — 규칙만 보존한다(부활 시 모순 방지).
 */
export function creatableRoles(actingRole: UserRole): readonly UserRole[] {
  switch (actingRole) {
    case "admin":
      return ["temp_user", "user", "advanced_user", "manager"];
    case "manager":
      return ["temp_user", "user", "advanced_user"];
    default:
      return [];
  }
}

export function canCreate(actingRole: UserRole, role: UserRole): boolean {
  return creatableRoles(actingRole).includes(role);
}
